core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -853,34 +853,37 @@ namespace kagome::blockchain {
 
   outcome::result<void> BlockTreeImpl::prune(
       const std::shared_ptr<TreeNode> &lastFinalizedNode) {
-    std::vector<std::pair<primitives::BlockHash, primitives::BlockNumber>>
-        to_remove;
-
-    auto current_node = lastFinalizedNode;
-
-    for (auto parent_node = current_node->parent.lock();
-         parent_node && !parent_node->finalized;
-         parent_node = current_node->parent.lock()) {
-      auto main_chain_node = current_node;
-      current_node = parent_node;
-
-      // collect hashes for removing (except main chain block)
-      for (const auto &child : current_node->children) {
-        if (child->block_hash != main_chain_node->block_hash) {
-          collectDescendants(child, to_remove);
-          to_remove.emplace_back(child->block_hash, child->depth);
+    std::list<std::shared_ptr<TreeNode>> to_remove;
+
+    auto following_node = lastFinalizedNode;
+
+    for (auto current_node = following_node->parent.lock();;) {
+      if (current_node) {
+        if (current_node->finalized && current_node->children.size() == 1) {
+          break;
+        }
+
+        // collect hashes for removing (except main chain block)
+        for (const auto &child : current_node->children) {
+          if (child->block_hash != following_node->block_hash) {
+            collectDescendants(child, to_remove);
+          }
         }
-      }
 
-      // remove (in memory) all child, except main chain block
-      current_node->children = {main_chain_node};
+        // remove (in memory) all child, except main chain block
+        current_node->children = {following_node};
+        following_node = current_node;
+        current_node = current_node->parent.lock();
+      } else {
+        break;
+      }
     }
 
     std::vector<primitives::Extrinsic> extrinsics;
 
     // remove from storage
-    for (const auto &[hash, number] : to_remove) {
-      auto block_body_res = storage_->getBlockBody(hash);
+    for (const auto &node : to_remove) {
```

Comment:
Why?

Reply:
To use is directly as a removeFromMeta argument - previously there was a need to search it again.